snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,256 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+from confluent_kafka import (
+    TIMESTAMP_LOG_APPEND_TIME,
+    Consumer as KafkaConsumer,
+    TopicPartition,
+)
+
+from snuba.utils.kafka.consumer import Consumer
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Position:
+    """
+    Represents a position in a Kafka partition.
+
+    This represents an offset, as well as a point in time.
+    """
+
+    offset: int
+    timestamp: float
+
+
+@dataclass(frozen=True)
+class Interval:
+    """
+    Represents a range between two positions in a Kafka partition.
+    """
+
+    lower: Position
+    upper: Position
+
+    def __post_init__(self) -> None:
+        assert self.upper.offset > self.lower.offset
+        assert self.upper.timestamp >= self.lower.timestamp
+
+    def shift(self, position: Position) -> Interval:
+        """
+        Return a new ``Interval`` instance, using the provided position as
+        the new upper bound, and the existing upper bound as the new lower
+        bound.
+        """
+        return Interval(self.upper, position)
+
+
+@dataclass(frozen=True)
+class Task:
+    """Represents a single task to be executed."""
```

Comment:
Details within this class are to be filled in, but ultimately this will have some mechanism for building a `Query` object to be executed.

Reply:
Makes sense, will do.

Reply:
Fixed with 6f7962ea0d95838717caff4167b06074cf703027.